src/lj_jitlog.c

```diff
@@ -0,0 +1,273 @@
+
+typedef struct JITLogState {
+  SBuf eventbuf; /* Must be first so loggers can reference it just by casting the G(L)->vmevent_data pointer */
+  JITLogUserContext user;
+  global_State *g;
+} JITLogState;
+
+
+static void free_context(JITLogState *context);
+
+static void jitlog_callback(void *contextptr, lua_State *L, int eventid, void *eventdata)
+{
+  VMEvent2 event = (VMEvent2)eventid;
+  JITLogState *context = contextptr;
+
+  switch (event) {
+    case VMEVENT_DETACH:
+      free_context(context);
+      break;
+    case VMEVENT_STATE_CLOSING:
+      if (G(L)->vmevent_cb == jitlog_callback) {
+        luaJIT_vmevent_sethook(L, NULL, NULL);
+      }
+      free_context(context);
+      break;
+    default:
+      break;
+  }
+}
+
+#if LJ_TARGET_X86ORX64
+
+static int getcpumodel(char *model)
+{
+  lj_vm_cpuid(0x80000002u, (uint32_t*)(model));
+  lj_vm_cpuid(0x80000003u, (uint32_t*)(model + 16));
+  lj_vm_cpuid(0x80000004u, (uint32_t*)(model + 32));
+  return (int)strnlen((char*)model, 12 * 4);
+}
+
+#else
+
+static int getcpumodel(char *model)
+{
+  strcpy(model, "unknown");
+  return (int)strlen("unknown");
+}
+
+#endif
+
+static int bufwrite_strlist(SBuf *sb, const char** list, int limit)
+{
+  const char** pos = list;
+  int count = 0;
+  for (; *pos != NULL && (limit == -1 || count < limit); pos++, count++) {
+    const char *s = *pos;
+    lj_buf_putmem(sb, s, (MSize)strlen(s)+1);
+  }
+  return count;
+}
+
+static void write_header(JITLogState *context)
+{
+  lua_State *L = mainthread(context->g);
+  char cpumodel[64] = {0};
+  int model_length = getcpumodel(cpumodel);
+  SBuf sb;
+  lj_buf_init(L, &sb);
+  bufwrite_strlist(&sb, msgnames, MSGTYPE_MAX);
+  log_header(context->g, 1, 0, sizeof(MSG_header), msgsizes, MSGTYPE_MAX, sbufB(&sb), sbuflen(&sb), cpumodel, model_length, LJ_OS_NAME, (uintptr_t)G2GG(context->g));
+  lj_buf_free(context->g, &sb);
+}
+
+static int jitlog_isrunning(lua_State *L)
+{
+  void* current_context = NULL;
+  luaJIT_vmevent_callback cb = luaJIT_vmevent_gethook(L, (void**)&current_context);
+  return cb == jitlog_callback;
+}
+
+LUA_API int luaopen_jitlog(lua_State *L);
+
+LUA_API JITLogUserContext* jitlog_start(lua_State *L)
+{
+  JITLogState *context ;
+  lua_assert(!jitlog_isrunning(L));
+
+  context = malloc(sizeof(JITLogState));
+  memset(context, 0 , sizeof(JITLogState));
+  context->g = G(L);
+
+  MSize total = G(L)->gc.total;
+  SBuf *sb = &context->eventbuf;
+  lj_buf_init(L, sb);
+  lj_buf_more(sb, 1024*1024*100);
+  /* 
+  ** Our Buffer size counts towards the gc heap size. This is an ugly hack to try and 
+  ** keep the GC running at the same rate with the jitlog running by excluding our buffer
+  ** from the gcheap size.
+  */
+  G(L)->gc.total = total;
+  luaJIT_vmevent_sethook(L, jitlog_callback, context);
+  write_header(context);
+
+  lj_lib_prereg(L, "jitlog", luaopen_jitlog, tabref(L->env));
+  return &context->user;
+}
+
+static int dumptofile(JITLogState *context, const char *path);
```

Comment:
it think this forward declaration, right before a function that is forward-declared itself, begs for some reordering.  the first one that comes to mind is to move `free_context()`, `jitlog_shutdown()` and `jitlog_close()` to just before `jlib_getstate()`.  Also to my eyes, the "closing" functions help to signal the end of a group of functions.  of course, i'd much better like those groups to be separated by comments (i.e. see those at lib_base.c)

Reply:
Well that function is half redundant really there should just be jitlog_save. For the ordering i like to keep the open and close functions together because there often touching the same variables it makes things easier when i make changes in one to keep it in sync with the other.

Reply:
yes, that's a matter of taste.  and about separating the sections with comments?

Reply:
Well i added a separator for Lua module but i could also add one at the start of the public api